File: day:28/hasSumPathTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        return hasSumFound(root, 0, sum);
    }
    
    public boolean hasSumFound(TreeNode node, int pval, int sum) {
        if(node == null) return false;
        var t = node.val + pval;
    
        if(node.left == null && node.right == null) {
            if(t == sum) {
                return true;
            }
            if( sum > 0 && t > sum) return false;
            if( sum < 0 && t < sum) return false;
        }
    
        var left = hasSumFound(node.left, t, sum);
        var right = hasSumFound(node.right, t, sum);
    
        return left || right;
    }
    
}

File: day:28/hasSumPathTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        return self.hasSumFound(root, 0, sum);
    
    def hasSumFound(self, node, pval, sum):
        if(node is None): return False;
        
        t = node.val + pval;
    
        if(node.left is None and node.right is None):
            if(t == sum):
                return True;
            if( sum > 0 and t > sum): return False;
            if( sum < 0 and t < sum): return False;
        left = self.hasSumFound(node.left, t, sum);
        right = self.hasSumFound(node.right, t, sum);
    
        return left or right; 
        
